### PR TITLE
De-deduplicate request types by extracting request type into separate interface

### PR DESCRIPTION
**Pull request by declanvong**
Output looks something like this now:
```ts
interface CourseWorkResource {

    ...

    patch(request: PatchCourseWorkResourceRequest): Request<CourseWork>;
    patch(
        request: Omit<PatchCourseWorkResourceRequest, 'resource'>,
        resource: PatchCourseWorkResourceRequest['resource'],
    ): Request<CourseWork>;
}

...

interface PatchCourseWorkResourceRequest {
    /** V1 error format. */
    "$.xgafv"?: string;
    /** OAuth access token. */
    access_token?: string;
    /** Data format for response. */
    alt?: string;

    ...

    /** Request body */
    resource: CourseWork;
}
```

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -240,6 +240,10 @@ class TypescriptTextWriter implements ITypescriptTextWriter {
   public method(name: string, parameters: { parameter: string, type: string | TypescriptWriterCallback }[], returnType: string, singleLine = false) {
     this.writer.startIndentedLine(`${name}(`);
 
+    if (!singleLine) {
```
Comment (declanvong) [inline src/app.ts]: Just some pretty-fying of the output, splits params onto newlines with trailing comma

File: src/app.ts [hunk 2]
```diff
@@ -382,14 +391,14 @@ function formatComment(comment: string) {
 }
 
 function getMethodReturn(method: gapi.client.discovery.RestMethod, schemas: Record<string, gapi.client.discovery.JsonSchema>) {
-
   const name = schemas['Request'] ? 'client.Request' : 'Request';
 
   if (method.response) {
-    const schema = schemas[checkExists(method.response.$ref)];
+    const schemaName = method.response.$ref;
```
Comment (declanvong) [inline src/app.ts]: (unrelated refactoring)

File: src/app.ts (hunk 1 above)
Comment (Maxim-Mazurok) [reply src/app.ts]: I think that we should give up any attempts to generate pretty output, because it'll mean that we should invent prettier. Lucky for us, it already exists :) I think we should generate ugly code and then just run it through prettier, based on DefinitelyTyped config.
Otherwise, LGTM :+1:

**Comment by declanvong** [comment]
Seems like this branch was merged into `declan/merge-declanvong-changes`, but when that branch was merged into master it didn't include this commit 🤔 
Will raise this one again to merge separately into master

**Comment by declanvong** [comment]
What does https://github.com/Maxim-Mazurok/google-api-typings-generator/commit/776e36ef25886fdb2d38a002ed12ba1dacde85c5 do? Just resolving the conflicts for it :)

**Comment by Maxim-Mazurok** [comment]
> What does [776e36e](https://github.com/Maxim-Mazurok/google-api-typings-generator/commit/776e36ef25886fdb2d38a002ed12ba1dacde85c5) do? Just resolving the conflicts for it :)

Some methods accept `resource` property. And some of them also accept request body.
We have to ways of passing request body: using `resource` property and using second argument (`body`). So in case if we have `resource` property and request body, we will generate two definitions:
```
x({
  resource: ActualResourceProperty,
  resource: RequestBody, // <-- duplicated resource, invalid
  other: prop
})
```
and
```
x({
  resource: ActualResourceProperty,
  other: prop
}, body: RequestBody)
```
You can see that first one is invalid, because it has two `resource` props.

That commit handles exactly this edge case, by generating only second variant (with `body` as a second argument).

**Comment by Maxim-Mazurok** [comment]
> Seems like this branch was merged into `declan/merge-declanvong-changes`, but when that branch was merged into master it didn't include this commit thinking
> Will raise this one again to merge separately into master

Oh, right, I think that I've merged `declan/merge-declanvong-changes` into `master` before merging this PR.. Didn't notice that it wasn't PR for `master` branch, sorry.

**Comment by declanvong** [comment]
Ahh, gotcha, cool will resolve the conflict and re-raise PR.

>Some methods accept resource property
why google why?!
